servers/rendering/shader_preprocessor.cpp

```diff
@@ -0,0 +1,1317 @@
+		return;
+	}
+
+	p_tokenizer->skip_whitespace();
+	if (!is_char_end(p_tokenizer->peek())) {
+		set_error(RTR("Invalid ifdef."), line);
+		return;
+	}
+	p_tokenizer->advance('\n');
+
+	bool success = state->defines.has(label);
+	start_branch_condition(p_tokenizer, success);
+}
+
+void ShaderPreprocessor::process_ifndef(PreprocessorTokenizer *p_tokenizer) {
+	const int line = p_tokenizer->get_line();
+
+	String label = p_tokenizer->get_identifier();
+	if (label.is_empty()) {
+		set_error(RTR("Invalid macro name."), line);
+		return;
+	}
+
+	p_tokenizer->skip_whitespace();
+	if (!is_char_end(p_tokenizer->peek())) {
+		set_error(RTR("Invalid ifndef."), line);
+		return;
+	}
+	p_tokenizer->advance('\n');
+
+	bool success = !state->defines.has(label);
+	start_branch_condition(p_tokenizer, success);
+}
+
+void ShaderPreprocessor::process_include(PreprocessorTokenizer *p_tokenizer) {
+	const int line = p_tokenizer->get_line();
+
+	p_tokenizer->advance('"');
+	String path = tokens_to_string(p_tokenizer->advance('"'));
+	path = path.substr(0, path.length() - 1);
+	p_tokenizer->skip_whitespace();
+
+	if (path.is_empty() || !is_char_end(p_tokenizer->peek())) {
+		set_error(RTR("Invalid path."), line);
+		return;
+	}
+
+	Ref<Resource> res = ResourceLoader::load(path);
+	if (res.is_null()) {
+		set_error(RTR("Shader include load failed. Does the shader exist? Is there a cyclic dependency?"), line);
+		return;
+	}
+
+	Ref<Shader> shader = Object::cast_to<Shader>(*res);
+	Ref<ShaderInclude> shader_inc = Object::cast_to<ShaderInclude>(*res);
+	if (shader.is_null() && shader_inc.is_null()) {
+		set_error(RTR("Shader include resource type is wrong."), line);
+		return;
+	}
+	bool is_shader = shader.is_valid();
+
+	String included = is_shader ? shader->get_code() : shader_inc->get_code();
+	if (!included.is_empty()) {
+		uint64_t code_hash = included.hash64();
+		if (state->cyclic_include_hashes.find(code_hash)) {
+			set_error(RTR("Cyclic include found."), line);
+			return;
+		}
+	}
+
+	if (is_shader) {
+		int type_end = included.find(";");
+		if (type_end == -1) {
+			set_error(RTR("Shader include `shader_type` not found."), line);
+			return;
+		}
+	}
+
+	const String real_path = is_shader ? shader->get_path() : shader_inc->get_path();
+	if (state->includes.has(real_path)) {
+		// Already included, skip.
+		// This is a valid check because 2 separate include paths could use some
+		// of the same shared functions from a common shader include.
+		return;
+	}
+
+	// Mark as included.
+	state->includes.insert(real_path);
+
+	state->include_depth++;
+	if (state->include_depth > 25) {
+		set_error(RTR("Shader max include depth exceeded."), line);
+		return;
+	}
+
+	String old_include = state->current_include;
+	state->current_include = real_path;
+	ShaderPreprocessor processor(included);
+
+	int prev_condition_depth = state->condition_depth;
+	state->condition_depth = 0;
+
+	String result;
+	processor.preprocess(state, result);
+	add_to_output(result);
```

Comment:
As I suggested, I think it would be much simpler that its the ShaderLanguage tokenizer that gets rid of correcting file and line by inserting special tokens for this. This way you can entirely discard the preprocessor after it did its job.

Reply:
Not clearly understand what you mean here and how it can be done. Can you post a sample code?